v0/destinations/clickup/util.js

```diff
@@ -0,0 +1,383 @@
+ * @param {*} properties {"url":"https://www.rudderstack.com/"}
+ * @param {*} listId
+ * @param {*} apiToken
+ * @returns [{"id":"b0f40a94-ea2a-4998-a514-8074d0eddcde","value":"https://www.rudderstack.com/"}]
+ */
+const customFieldsBuilder = async (
+  keyToCustomFieldName,
+  properties,
+  listId,
+  apiToken
+) => {
+  const responseArray = [];
+  if (properties && keyToCustomFieldName) {
+    // retrieve available clickup custom field for the given list
+    const availableCustomFields = await retrieveCustomFields(listId, apiToken);
+    // convert array to hashMap with key as field name and value as custom field object
+    const availableCustomFieldsMap = getHashFromArrayWithValueAsObject(
+      availableCustomFields,
+      "name",
+      false
+    );
+
+    // convert destination.Config.keyToCustomFieldName to hashMap
+    const configCustomFieldsMap = getHashFromArray(
+      keyToCustomFieldName,
+      "from",
+      "to",
+      false
+    );
+
+    Object.keys(configCustomFieldsMap).forEach(propertiesKey => {
+      let fieldValue = properties[propertiesKey];
+      if (fieldValue) {
+        const fieldName = configCustomFieldsMap[propertiesKey];
+        const customFieldDetailsObject = availableCustomFieldsMap[fieldName];
+        validateUserSentCustomFieldValues(customFieldDetailsObject, fieldValue);
+        fieldValue = populateCustomFieldValue(
+          customFieldDetailsObject,
+          fieldValue
+        );
+        if (fieldValue && customFieldDetailsObject) {
+          responseArray.push({
+            id: customFieldDetailsObject.id,
+            value: fieldValue
+          });
+        }
+      }
+    });
+  }
+  return responseArray;
+};
+
+/**
+ * Function to discard any event that is not a part of destination.Config.whitelistedEvents
+ * @param {*} message
+ * @param {*} destination
+ */
+const checkEventIfUIMapped = (message, destination) => {
+  const { whitelistedEvents } = destination.Config;
+  const { event } = message;
+  const nonEmptyWhitelistedEvents = whitelistedEvents?.filter(
+    whiteListedEvent => whiteListedEvent.eventName
+  );
+
+  if (nonEmptyWhitelistedEvents && nonEmptyWhitelistedEvents.length > 0) {
+    const allowEvent = nonEmptyWhitelistedEvents.some(
+      whiteListedEvent =>
+        whiteListedEvent.eventName.toLowerCase() === event.toLowerCase()
+    );
+    if (!allowEvent) {
+      throw new CustomError(
+        "[ CLICKUP ]:: Event Discarded. To allow this event, add this in Allowlist",
+        400
+      );
+    }
+  }
+};
+
+const isDefinedAndNotNullAndNotEmpty = value => {
+  return !(
+    value === undefined ||
+    value === null ||
+    Number.isNaN(value) ||
+    (typeof value === "object" && Object.keys(value).length === 0) ||
+    (typeof value === "string" && value.trim().length === 0)
+  );
+};
+
+const removeUndefinedAndNullAndEmptyValues = obj =>
+  _.pickBy(obj, isDefinedAndNotNullAndNotEmpty);
```

Comment:
Move this to global util file

Reply:
Function with the same name is present in global utils. Should I do the function versioning here?

Reply:
Function moved to global utils.